File: crud-item-list/src/main/java/com/gurunars/crud_item_list/Action.java
package com.gurunars.crud_item_list;

import java.util.List;
import java.util.Set;

interface Action<ItemType> {

    boolean perform(List<ItemType> all, Set<ItemType> selectedItems);
    boolean canPerform(List<ItemType> all, Set<ItemType> selectedItems);

}

File: crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/Action.kt
package com.gurunars.crud_item_list

import com.gurunars.box.IRoBox
import com.gurunars.item_list.Item

internal typealias ItemSetChange<ItemType> =
    (newAll: List<ItemType>, newSelectedItems: Set<ItemType>) -> Unit

internal interface Action<ItemType : Item> {
    fun perform(
        all: List<ItemType>,
        selectedItems: Set<ItemType>,
        consumer: ItemSetChange<ItemType>
    )
    fun canPerform(
        all: IRoBox<List<ItemType>>,
        selectedItems: IRoBox<Set<ItemType>>
    ): IRoBox<Boolean>
}
